ARROW-1572: [C++] Implement "value counts" kernels for tabulating value frequencies

emkornfield opened the pull request:
Picked up from: https://github.com/apache/arrow/pull/1970

I mostly reused the unit tests as is and modified the rest based on feedback on that PR and adapting to new code.  Some other changes I mae:
1.  Remove the HashKernel and getter from the public API.  I think we can add these back once we have a better idea on how we are doing stateful kernels (i.e. https://github.com/apache/arrow/pull/3407)
2. Add operator[] to NumericBuilder to modify previously added values (this might not be the right place, and I had a little bit of trouble figuring out how to integrate this into the existing TypedTest so the testing is a little bit weak).  This seemed better then using a vector to collect values.

If this approach looks OK for now.  I'm also going to open up a JIRA to try refactor the unittest for Hash kernels (and maybe the headers) since I think there might be a clearer more granular way of laying these out.

other things to discuss:
1.  Handling null values in Count/Unique (it looks like they are dropped today, should this configurable/turned on).
2.  Hashing edge cases for floating point numbers (just opened a JIRA on this).

andrioni commented:
Thanks for picking this up from where I left off, I'll close my original PR.

emkornfield commented:
@andrioni Thanks for the contribution

fsaintjacques commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -72,19 +72,69 @@ class UniqueAction {
+  Status FlushFinal(Datum* out) {
+    std::shared_ptr<ArrayData> result;
+    RETURN_NOT_OK(count_builder_.FinishInternal(&result));
+    out->value = std::move(result);
```
For completeness, you should set the datum type if we ever refactor.

emkornfield replied:
Could you clarify?  I might be missing it but I don't see an API to set the type explicitly (instead it looks like it gets derived when value that is assigned in the underlying variant)

wesm replied:
It's handled by the variant

fsaintjacques replied:
Yes, my bad.

xhochy commented:
Rebased

emkornfield commented:
@fsaintjacques thanks for the review.  My thoughts:

* May I propose to rename this to `Distribution`?
** I renamed to ValueCounts to match up with corresponding Pandas function (https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.value_counts.html), Distribution might also make sense.  Since naming is hard, I'd like to get a second opinion  before changing.
* We should also count Nulls.
** I think this should be configurable, similar to Pandas, (and affects Uniques) as well, since this is a bigger change, I'd like to handle under a different JIRA if that is OK.
* The current output format is 2 arrays. I think we should formalize this and return a single array of `List<Struct<Option<value>, count>>` explicitly which should allow returning the count for the null value.
** Agreed, I am working through making this happen and will update the PR when I complete it.

emkornfield commented on cpp/src/arrow/compute/kernels/hash.h:
```diff
@@ -66,6 +47,24 @@ Status GetDictionaryEncodeKernel(FunctionContext* ctx,
+/// \note API not yet finalized
+ARROW_EXPORT
+Status ValueCounts(FunctionContext* context, const Datum& value,
+                   std::shared_ptr<Array>* counts);
```
this could potentially return a StructArray directly.

emkornfield commented:
@fsaintjacques I think I fixed the windows build (will check in the morning).  I changed the output to be a Struct as suggested.  I will file follow-up JIRAs for the rest.

fsaintjacques commented on cpp/src/arrow/util/hashing.h:
```diff
@@ -473,7 +473,9 @@ class SmallScalarMemoTable {
   // Copy values starting from index `start` into `out_data`
   void CopyValues(int32_t start, Scalar* out_data) const {
-    memcpy(out_data, &index_to_value_[start], size() - start);
+    if (index_to_value_.size() > 0) {
```
I'd put a DCHECK_GE(start, 0)` just to make sure it's positive.

emkornfield replied:
Also put a ASSERT_LE in and changed to std::copy which i think fixes another bug.

emkornfield commented:
Clean build 👍

fsaintjacques commented:
I want to give a last review, but I don't think anything is blocking.

fsaintjacques commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -415,5 +506,27 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
+  RETURN_NOT_OK(func->FlushFinal(&value_counts));
+
+  auto data_type = std::make_shared<StructType>(std::vector<std::shared_ptr<Field>>{
+      std::make_shared<Field>("Values", uniques->type()),
```
Make `Values` and `Count` a public `constexpr` such that user can programatically retrieve the field by name at compile time, maybe expose the index that way too.

emkornfield replied:
Good suggestion.  Done.

fsaintjacques commented:
Ready to merge!

fsaintjacques commented:
I think you need to rebase for the failing appveyor failure, but I'd say this is safe to merge. LGTM +1.

emkornfield commented:
@wesm or @xhochy mind taking a look to merge.  I believe the build failure was spurious due to https://github.com/apache/arrow/pull/3821

wesm commented on cpp/src/arrow/compute/kernels/hash-test.cc:
```diff
@@ -61,9 +61,46 @@ void CheckUnique(FunctionContext* ctx, const shared_ptr<DataType>& type,
+      ArrayData::Make(type, 0 /* length */, std::move(data_buffers), 0 /* null_count */);
+
+  shared_ptr<Array> ex_values = _MakeArray<Type, T>(type, {}, {});
+  shared_ptr<Array> ex_counts = _MakeArray<Int64Type, int64_t>(int64(), {}, {});
```
Could also use `ArrayFromJSON` here

wesm commented on cpp/src/arrow/compute/kernels/hash-test.cc:
```diff
@@ -329,6 +451,14 @@ TEST_F(TestHashKernel, ChunkedArrayInvoke) {
 
+  // Unique counts
+  shared_ptr<Array> counts_array;
+  ASSERT_OK(ValueCounts(&this->ctx_, Datum(carr), &counts_array));
```
NB: not necessary to write `Datum(...)` here because of implicit ctor

wesm commented on cpp/src/arrow/compute/kernels/hash-test.cc:
```diff
@@ -329,6 +451,14 @@ TEST_F(TestHashKernel, ChunkedArrayInvoke) {
+  ASSERT_OK(ValueCounts(&this->ctx_, Datum(carr), &counts_array));
+  auto counts_struct = std::dynamic_pointer_cast<StructArray>(counts_array);
+  ASSERT_ARRAYS_EQUAL(*ex_dict, *counts_struct->field(0));
+  ASSERT_ARRAYS_EQUAL(*ex_counts, *counts_struct->field(1));
```
It seems reasonable to expect a consistent field position

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -82,21 +82,77 @@ class UniqueAction : public ActionBase {
+    return Status::OK();
+  }
+
+  // Don't do anything on flush beceause we don't want to finalize the builder
```
typo

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -82,21 +82,77 @@ class UniqueAction : public ActionBase {
+    return Status::OK();
+  }
+
+  Status ObserveNull() { return Status::OK(); }
```
We're going to want to count nulls (perhaps optionally). The null count can be appended at the end

https://issues.apache.org/jira/browse/ARROW-4787

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -82,21 +82,77 @@ class UniqueAction : public ActionBase {
+
+  template <class Index>
+  Status ObserveFound(Index slot) {
+    count_builder_[slot]++;
```
I'm curious how efficient the code that this generates is

emkornfield replied:
In release things seem to get inlined (can't find ObserveFound anymore).  In debug:
```
0000000000513740 <_ZN5arrow7compute12_GLOBAL__N_117ValueCountsAction12ObserveFoundIiEEvT_>:
  513740:       55                      push   %rbp
  513741:       48 89 e5                mov    %rsp,%rbp
  513744:       48 83 ec 10             sub    $0x10,%rsp
  513748:       48 89 7d f8             mov    %rdi,-0x8(%rbp)
  51374c:       89 75 f4                mov    %esi,-0xc(%rbp)
  51374f:       48 8b 7d f8             mov    -0x8(%rbp),%rdi
  513753:       48 83 c7 18             add    $0x18,%rdi
  513757:       48 63 75 f4             movslq -0xc(%rbp),%rsi
  51375b:       e8 90 1f d5 ff          callq  2656f0 <_ZN5arrow14NumericBuilderINS_9Int64TypeEEixEl@plt>
  513760:       48 8b 30                mov    (%rax),%rsi
  513763:       48 83 c6 01             add    $0x1,%rsi
  513767:       48 89 30                mov    %rsi,(%rax)
  51376a:       48 83 c4 10             add    $0x10,%rsp
  51376e:       5d                      pop    %rbp
  51376f:       c3                      retq   
```
```
00000000002656f0 <_ZN5arrow14NumericBuilderINS_9Int64TypeEEixEl@plt>:
  2656f0:       ff 25 8a 24 9f 00       jmpq   *0x9f248a(%rip)        # c57b80 <_ZN5arrow14NumericBuilderINS_9Int64TypeEEixEl@@Base+0x731e70>
  2656f6:       68 6d 13 00 00          pushq  $0x136d
  2656fb:       e9 10 c9 fe ff          jmpq   252010 <.plt>
```

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -110,15 +166,19 @@ class DictEncodeAction : public ActionBase {
-  void ObserveFound(Index index) {
+  Status ObserveFound(Index index) {
     indices_builder_.UnsafeAppend(index);
+    return Status::OK();
```
I wonder if this has performance implications. I'll run the perf benchmarks before/after to see

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -278,44 +363,45 @@ struct HashKernelTraits<Type, Action, enable_if_fixed_size_binary<Type>> {
 
 }  // namespace
 
+#define PROCESS_SUPPORTED_HASH_TYPES \
```
I often pass `PROCESS` in as an argument to be more explicit

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -415,5 +506,31 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
 }
 
+const char kValuesFieldName[] = "Values";
+const char kCountsFieldName[] = "Counts";
```
I have a personal preference for lowercase, but I don't think it's too important for now

wesm commented on cpp/src/arrow/compute/kernels/hash.h:
```diff
@@ -34,29 +34,10 @@ namespace compute {
 
-/// \brief Invoke hash table kernel on input array, returning any output
-/// values. Implementations should be thread-safe
-class ARROW_EXPORT HashKernel : public UnaryKernel {
```
I think we _will_ want to publicly expose this at some point for parallelization. But I suppose it is OK to leave it out of the public header for now

emkornfield replied:
Agreed, I didn't want to sign-up for a contract to this until a little more thought it given.  We should probably reorganize the code a little bit in general (I don't think everything having to do with hashing belongs in this header and .cc file).  I think most of the other kernel implementations are also private.

wesm commented on cpp/src/arrow/compute/kernels/hash.cc:
```diff
@@ -415,5 +506,31 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
+      std::make_shared<Field>("Counts", int64())});
+  *counts = std::make_shared<StructArray>(
+      data_type, uniques->length(),
+      std::vector<std::shared_ptr<Array>>{uniques, MakeArray(value_counts.array())});
```
We should come up with a more elegant API for constructing StructArray https://issues.apache.org/jira/browse/ARROW-4788

wesm commented on cpp/src/arrow/compute/kernels/hash.h:
```diff
@@ -66,6 +47,29 @@ Status GetDictionaryEncodeKernel(FunctionContext* ctx,
+ARROW_EXPORT extern const char kValuesFieldName[];
+ARROW_EXPORT extern const char kCountsFieldName[];
+ARROW_EXPORT extern const int32_t kValuesFieldIndex;
+ARROW_EXPORT extern const int32_t kCountsFieldIndex;
```
Hm, I'm having a YAGNI reaction to this. Premature development?

emkornfield replied:
For all of them?  I think we should probably provide one or another so there isn't ambiguity?

wesm commented:
Welp, bad news. The performance regression makes this patch unmergeable for now

master branch

```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       2337 us       2337 us        617   1.66998GB/s
BM_BuildStringDictionary/min_time:1.000                                 3947 us       3947 us        358   76.4939MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             30889 us      30888 us         45   4.04681GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           28470 us      28469 us         48   4.39063GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          45673 us      45672 us         31   2.73684GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time       424376 us     424374 us          3   301.619MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           44474 us      44474 us         31   2.81063GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         38880 us      38879 us         35   3.21505GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time        55346 us      55346 us         25   2.25852GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     393982 us     393977 us          4   324.888MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           217200 us     217199 us          7   736.649MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         184280 us     184279 us          7   868.243MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        298878 us     298877 us          5   535.335MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     1895167 us    1895146 us          1   84.4253MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          324695 us     324692 us          4   4.81221GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        407262 us     407258 us          3    3.8366GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time       579285 us     579277 us          2   2.69729GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    2684019 us    2683992 us          1   596.121MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time             9999 us       9998 us        140   1.56273GB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          20517 us      20516 us         68   779.853MB/s
```

this PR

```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       2964 us       2964 us        486   1.31671GB/s
BM_BuildStringDictionary/min_time:1.000                                 4740 us       4740 us        296   63.6928MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             58138 us      58137 us         24   2.15007GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           53214 us      53213 us         25   2.34903GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          73082 us      73081 us         19   1.71041GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time       518587 us     518583 us          3   246.824MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           79186 us      79185 us         18   1.57857GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         69846 us      69845 us         20   1.78966GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time        86503 us      86502 us         16   1.44504GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     477864 us     477860 us          3   267.859MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           265399 us     265396 us          5   602.865MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         231742 us     231739 us          6   690.422MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        350534 us     350530 us          4   456.447MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     1891871 us    1891853 us          1   84.5723MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          378638 us     378635 us          4   4.12664GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        461684 us     461680 us          3   3.38435GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time       637809 us     637804 us          2   2.44979GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    2997666 us    2997611 us          1   533.749MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            39592 us      39592 us         35    404.12MB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          53375 us      53375 us         26   299.764MB/s
```

wesm commented:
By the way, I think it's growing less and less responsible by the day for us to continue to merge patches to master without running automated perf regression tests. @kszucs what do you think the road is like to set up a bot that can do automated reporting on major perf changes on PRs?

wesm commented:
cc @aregm

emkornfield commented:
@wesm thanks for review and running the benchmark, I should have done it myself.  I will try to have a new version with hopefully equivalent performances numbers by sometime next week.

emkornfield commented:
Still looks slower (but not in all cases and my machine might not be the best for benchmarking):
```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       3410 us       3383 us        423   1.15354GB/s
BM_BuildStringDictionary/min_time:1.000                                 7907 us       7875 us        183   38.3358MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             50963 us      50829 us         29   2.45274GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           62802 us      62032 us         28   1.99039GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          81648 us      81426 us         16   1.53096GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time      1054502 us    1050954 us          1   121.384MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           74412 us      74220 us         19   1.67984GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         70355 us      69901 us         21    1.7767GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time       131523 us     131131 us         10   973.214MB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time    1079100 us    1075032 us          1   118.617MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           346633 us     345537 us          4   461.583MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         301914 us     301341 us          4   529.953MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        471556 us     469987 us          3   339.302MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     3576311 us    3564658 us          1   44.7388MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          510885 us     509112 us          3   3.05842GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        604308 us     602899 us          2    2.5856GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time      1027190 us    1023880 us          1   1.52114GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    5666492 us    5647275 us          1   282.362MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            64508 us      64342 us         22   248.031MB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          79491 us      79240 us         18    201.28MB/s
```
This PR:
```
With changes.
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       3312 us       3302 us        439     1.182GB/s
BM_BuildStringDictionary/min_time:1.000                                 7916 us       7897 us        181   38.2297MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             52459 us      52332 us         26   2.38281GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           74160 us      73440 us         28   1.68554GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time         101905 us     101462 us         13   1.22663GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time      1486438 us    1480525 us          1   86.1119MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           74015 us      73796 us         19   1.68884GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         69954 us      69725 us         21   1.78689GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time       102017 us     101761 us         15   1.22529GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     853931 us     851550 us          2   149.895MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           330582 us     329818 us          4   483.994MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         309359 us     308100 us          5   517.198MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        630798 us     628967 us          3   253.647MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     5302174 us    5284662 us          1   30.1763MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          546463 us     544996 us          2    2.8593GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        637016 us     635738 us          2   2.45284GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time      1053228 us    1050351 us          1   1.48354GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    5045640 us    5029777 us          1   317.105MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            60625 us      60487 us         22   263.918MB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          76488 us      76244 us         19   209.184MB/s
```

emkornfield commented:
I'm going to try switching back to memcpy to see if that has an affect.

emkornfield commented:
With change back to memcpy:
```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       3215 us       3206 us        468    1.2174GB/s
BM_BuildStringDictionary/min_time:1.000                                 7390 us       7366 us        189    40.986MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             47985 us      47844 us         29   2.60501GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           49667 us      49491 us         28   2.51674GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          79659 us      79425 us         17    1.5692GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time      1448104 us    1438814 us          1   88.3914MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           71486 us      71274 us         14   1.74859GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         75702 us      75453 us         20    1.6512GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time       115880 us     115201 us         14   1104.59MB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     793041 us     791101 us          2   161.404MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           321899 us     320826 us          4    497.05MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         295318 us     294576 us          5   541.788MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        467910 us     466879 us          3   341.946MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     3271504 us    3239283 us          1   48.9072MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          462968 us     461677 us          3   3.37496GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        609538 us     607450 us          2   2.56342GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time      1090298 us    1086588 us          1   1.43309GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    5549959 us    5529844 us          1    288.29MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            58626 us      58514 us         24   272.916MB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          73221 us      73044 us         19   218.517MB/s
```

emkornfield commented:
@wesm  this still has at least one large diff, but most number on my virt look better.  Assuming the build is green (i.e. I didn't make a silly mistake someplace).  Could you rerun on your box to see if you get the same results.  If this isn't an acceptable result, I would be open to suggestions for other things to try (off the top of my mind I can only think of having two copies of the hash code to make sure things are untouched).

fsaintjacques commented:
30% slower in the worst case.

![benchmark](https://user-images.githubusercontent.com/662469/54133762-044d5d80-43ed-11e9-90b0-1f4751caa64a.png)

Textual ouput:

```
01:cpp/ (pr/3579) ! ~/src/benchmark/tools/compare.py benchmarks build/release/arrow-compute-benchmark build-pr/release/arrow-compute-benchmark                                                                                                                                                                       [10/1881]
RUNNING: build/release/arrow-compute-benchmark --benchmark_out=/tmp/tmpc80tl0c1
2019-03-11 10:55:18                                                                                      
Running build/release/arrow-compute-benchmark                                                           
Run on (16 X 4500 MHz CPU s)                                                                             
CPU Caches:                                                                                                           
  L1 Data 32K (x8)                                                                                                    
  L1 Instruction 32K (x8)                                                                                             
  L2 Unified 1024K (x8)                                                                                               
  L3 Unified 11264K (x1)                                                                                              
---------------------------------------------------------------------------------------------------------             
Benchmark                                                                  Time           CPU Iterations              
---------------------------------------------------------------------------------------------------------             
BM_BuildDictionary/min_time:1.000                                       1949 us       1949 us        717    2.0024GB/s
BM_BuildStringDictionary/min_time:1.000                                 4399 us       4399 us        318   68.6282MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             31416 us      31415 us         44   3.97885GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           28285 us      28284 us         49   4.41933GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          32902 us      32900 us         41   3.79915GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time       502203 us     502198 us          3   254.877MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           43896 us      43896 us         32   2.84762GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         42571 us      42569 us         33   2.93628GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time        45076 us      45074 us         30   2.77311GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     485597 us     485587 us          3   263.593MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           202364 us     202354 us          7   790.656MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         167419 us     167409 us          8   955.686MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        245131 us     245126 us          6   652.714MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     2271991 us    2271905 us          1   70.4228MB/s                                          
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          332897 us     332880 us          4   4.69365GB/s                                          
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        379102 us     379086 us          4   4.12158GB/s                                          
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time       718619 us     718576 us          2   2.17431GB/s                                          
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    3645699 us    3645592 us          1   438.873MB/s                                          
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time             9091 us       9091 us        154   1.71869GB/s                                          
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          18952 us      18951 us         74   844.243MB/s                                          
RUNNING: build-pr/release/arrow-compute-benchmark --benchmark_out=/tmp/tmp31_5vpe3                                                                              
2019-03-11 10:56:22                                                                                                                                             
Running build-pr/release/arrow-compute-benchmark                                                                                                                
Run on (16 X 4500 MHz CPU s)                                                                                                                                    
CPU Caches:                                                                                                                                                     
  L1 Data 32K (x8)                                                                                                                                              
  L1 Instruction 32K (x8)                                                                                                                                       
  L2 Unified 1024K (x8)                                                                                                                                         
  L3 Unified 11264K (x1)                                                                                                                                        
---------------------------------------------------------------------------------------------------------                                                       
Benchmark                                                                  Time           CPU Iterations                                                        
---------------------------------------------------------------------------------------------------------                                                       
BM_BuildDictionary/min_time:1.000                                       2002 us       2002 us        695   1.94884GB/s                                          
BM_BuildStringDictionary/min_time:1.000                                 4281 us       4281 us        326   70.5266MB/s                                          
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             33747 us      33746 us         42   3.70408GB/s                                          
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           30347 us      30347 us         46   4.11902GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          37250 us      37249 us         36   3.35571GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time       508565 us     508551 us          3   251.689MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           46735 us      46734 us         30   2.67465GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         42668 us      42668 us         33   2.92958GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time        46289 us      46288 us         30   2.70041GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     482773 us     482768 us          3   265.135MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           205395 us     205390 us          7   778.988MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         172027 us     172022 us          8   930.085MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        248032 us     248029 us          6   645.078MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     2291535 us    2291492 us          1   69.8222MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          313713 us     313709 us          4   4.98066GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        381982 us     381974 us          4   4.09051GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time       720718 us     720694 us          2   2.16798GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    3819169 us    3819099 us          1   418.939MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            11793 us      11793 us        119   1.32493GB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          22288 us      22288 us         63    717.86MB/s
Comparing build/release/arrow-compute-benchmark to build-pr/release/arrow-compute-benchmark
Benchmark                                                                           Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                                +0.0275         +0.0275          1949          2002          1949          2002
BM_BuildStringDictionary/min_time:1.000                                          -0.0269         -0.0269          4399          4281          4399          4281
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time                       +0.0742         +0.0742         31416         33747         31415         33746
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time                     +0.0729         +0.0729         28285         30347         28284         30347
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time                    +0.1321         +0.1322         32902         37250         32900         37249
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time                  +0.0127         +0.0127        502203        508565        502198        508551
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time                     +0.0647         +0.0647         43896         46735         43896         46734
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time                   +0.0023         +0.0023         42571         42668         42569         42668
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time                  +0.0269         +0.0269         45076         46289         45074         46288
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time                -0.0058         -0.0058        485597        482773        485587        482768
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time                      +0.0150         +0.0150        202364        205395        202354        205390
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time                    +0.0275         +0.0276        167419        172027        167409        172022
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time                   +0.0118         +0.0118        245131        248032        245126        248029
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time                 +0.0086         +0.0086       2271991       2291535       2271905       2291492
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time                     -0.0576         -0.0576        332897        313713        332880        313709
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time                   +0.0076         +0.0076        379102        381982        379086        381974
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time                  +0.0029         +0.0029        718619        720718        718576        720694
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time                +0.0476         +0.0476       3645699       3819169       3645592       3819099
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time                      +0.2972         +0.2972          9091         11793          9091         11793
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time                    +0.1761         +0.1761         18952         22288         18951         22288

wesm commented:
Any thoughts about how to automate benchmark comparisons (Dockerized?)? It's currently a by-hand affair and potentially error prone

xhochy commented:
> Any thoughts about how to automate benchmark comparisons (Dockerized?)? It's currently a by-hand affair and potentially error prone

We could probably add a simple entry to `docker-compose.yml` that runs a different script in the `cpp` container that executes the benchmarks instead of the unittests. Can you open an issue?

wesm commented:
https://issues.apache.org/jira/browse/ARROW-4827

emkornfield commented:
One thing that I'll do which I'm not sure will help too much is separate out the memcpy fix into it's own PR so we can isolate performance problems to the kernel

emkornfield commented:
I think there might also be a way forward with template magic to eliminate this altogether.  I'll get an update out soon

emkornfield commented:
I think there might also be a way forward with template magic to eliminate this altogether.  I'll get an update out soon

emkornfield commented:
I think there might also be a way forward with template magic to eliminate this altogether.  I'll get an update out soon

emkornfield commented:
With this change it looks like they are fairly close on my box (rerunning causes changes in different directions in different metrics)

<img width="1090" alt="Screen Shot 2019-03-11 at 9 24 19 PM" src="https://user-images.githubusercontent.com/17869838/54174542-38a03880-4444-11e9-85fe-3ee9b6275d14.png">
Raw text:
```
Benchmark                                                                    Time             CPU      Time Old      Time New       CPU Old       CPU New
---------------------------------------------------------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                         -0.0038         -0.0036          3050          3039          3041          3030
BM_BuildStringDictionary/min_time:1.000                                   -0.0362         -0.0310          7278          7014          7211          6987
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time                +0.0020         +0.0006         47799         47894         47697         47724
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time              +0.0125         +0.0172         47223         47813         46874         47678
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time             +0.0112         +0.0110         81461         82375         81277         82175
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time           -0.0059         -0.0065        959365        953714        956518        950281
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time              +0.0295         +0.0304         68487         70507         68319         70399
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time            +0.0095         +0.0092         65062         65680         64903         65499
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time           +0.0364         +0.0366         88717         91943         88515         91752
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time         +0.0091         +0.0091        775562        782628        773094        780125
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time               -0.0001         -0.0007        325153        325115        324190        323967
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time             -0.0109         -0.0039        275206        272213        272525        271460
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time            -0.0373         -0.0374        458229        441117        456930        439855
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time          +0.0183         +0.0182       3150635       3208440       3142222       3199298
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time              -0.0034         -0.0022        465118        463514        463705        462684
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time            -0.0381         -0.0383        644111        619588        642393        617816
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time           +0.0546         +0.0554        957323       1009574        954481       1007365
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time         +0.0170         +0.0186       5132796       5220195       5099248       5193933
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time               +0.0063         +0.0070         60059         60437         59915         60334
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time             -0.0231         -0.0229         74762         73035         74568         72859
```

wesm commented:
Looking at this again and then merging, thanks for your patience

wesm commented:
I ran on a mobile Xeon 2.8 ghz with the cores pinned at 2.8ghz / no turbo boost and things looking OK here

```
before

```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       2950 us       2950 us        476   1.32273GB/s
BM_BuildStringDictionary/min_time:1.000                                 5941 us       5940 us        234   50.8258MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             46613 us      46614 us         30   2.68166GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           44150 us      44142 us         27   2.83129GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          65805 us      65805 us         21   1.89955GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time       505750 us     505756 us          3    253.09MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           70311 us      70312 us         20   1.77782GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         62900 us      62901 us         22   1.98729GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time        85752 us      85754 us         17   1.45769GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     486725 us     486734 us          3   262.982MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           308199 us     308202 us          4   519.145MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         262568 us     262572 us          5   609.367MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        439183 us     439189 us          3   364.313MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     2329791 us    2329828 us          1   68.6757MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          693907 us     693920 us          2   2.25174GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        821354 us     821368 us          2   1.90235GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time      1094868 us    1094886 us          1   1.42711GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    3770885 us    3770950 us          1   424.304MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            13690 us      13690 us        102   1.14134GB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          31576 us      31574 us         45   506.714MB/s
```

after

```
---------------------------------------------------------------------------------------------------------
Benchmark                                                                  Time           CPU Iterations
---------------------------------------------------------------------------------------------------------
BM_BuildDictionary/min_time:1.000                                       2969 us       2969 us        472   1.31424GB/s
BM_BuildStringDictionary/min_time:1.000                                 5966 us       5966 us        235   50.6048MB/s
BM_UniqueInt64NoNulls/16777216/50/min_time:1.000/real_time             46763 us      46763 us         30   2.67303GB/s
BM_UniqueInt64NoNulls/16777216/1024/min_time:1.000/real_time           44425 us      44426 us         32   2.81373GB/s
BM_UniqueInt64NoNulls/16777216/10240/min_time:1.000/real_time          69868 us      69869 us         21    1.7891GB/s
BM_UniqueInt64NoNulls/16777216/1048576/min_time:1.000/real_time       511248 us     511255 us          3   250.368MB/s
BM_UniqueInt64WithNulls/16777216/50/min_time:1.000/real_time           70771 us      70770 us         20   1.76626GB/s
BM_UniqueInt64WithNulls/16777216/1024/min_time:1.000/real_time         63143 us      63143 us         22   1.97964GB/s
BM_UniqueInt64WithNulls/16777216/10240/min_time:1.000/real_time        83603 us      83604 us         17   1.49516GB/s
BM_UniqueInt64WithNulls/16777216/1048576/min_time:1.000/real_time     486510 us     486505 us          3   263.098MB/s
BM_UniqueString10bytes/16777216/50/min_time:1.000/real_time           305595 us     305600 us          5   523.569MB/s
BM_UniqueString10bytes/16777216/1024/min_time:1.000/real_time         251144 us     251149 us          6   637.084MB/s
BM_UniqueString10bytes/16777216/10240/min_time:1.000/real_time        428113 us     428118 us          3   373.733MB/s
BM_UniqueString10bytes/16777216/1048576/min_time:1.000/real_time     2338876 us    2338908 us          1   68.4089MB/s
BM_UniqueString100bytes/16777216/50/min_time:1.000/real_time          691089 us     691102 us          2   2.26093GB/s
BM_UniqueString100bytes/16777216/1024/min_time:1.000/real_time        803811 us     803825 us          2   1.94387GB/s
BM_UniqueString100bytes/16777216/10240/min_time:1.000/real_time      1103679 us    1103689 us          1   1.41572GB/s
BM_UniqueString100bytes/16777216/1048576/min_time:1.000/real_time    4039441 us    4039457 us          1   396.094MB/s
BM_UniqueUInt8NoNulls/16777216/200/min_time:1.000/real_time            13894 us      13894 us        101   1.12461GB/s
BM_UniqueUInt8WithNulls/16777216/200/min_time:1.000/real_time          30398 us      30398 us         46   526.348MB/s
```

merging. thanks @emkornfield!
